flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java

```diff
@@ -435,4 +493,72 @@ private boolean isEndOfStreamReached(T record) {
             return isStreamEnd;
         }
     }
+
+    /**
+     * A wrapper around {@link SourceOutput} that counts the number of records emitted.
+     *
+     * <p>This wrapper is used when rate limiting is enabled to track how many records have been
+     * emitted since the last rate limit check, allowing the reader to properly apply backpressure
+     * when the rate limit is exceeded.
+     *
+     * @param <T> The type of records being emitted
+     */
+    private static final class RecordCountingSourceOutputWrapper<T> implements SourceOutput<T> {
```

Comment:
We can rename RecordCountingSourceOutputWrapper => RateLimitingSourceOutputWrapper, which is more Intuitively？